src/backend/functions/projects-edit.ts

```diff
@@ -0,0 +1,492 @@
+const addDescriptionBullets = async (
+  addedDetails: string[],
+  id: number,
+  descriptionBulletIdField: string
+) => {
+  // add the added bullets
+  if (addedDetails.length > 0) {
+    await prisma.description_Bullet.createMany({
+      data: addedDetails.map((element) => {
+        return {
+          detail: element,
+          [descriptionBulletIdField]: id
+        };
+      })
+    });
+  }
+};
+
+// edit descrption bullets in the db for each id and detail pair
+const editDescriptionBullets = async (editedIdsAndDetails: { id: number; detail: string }[]) => {
+  // edit the edited bullets if there are any to update
+  if (editedIdsAndDetails.length > 0) {
+    await prisma.$transaction(
+      editedIdsAndDetails.map((element) =>
+        prisma.description_Bullet.update({
+          where: {
+            descriptionId: element.id
+          },
+          data: {
+            detail: element.detail
+          }
+        })
+      )
+    );
+  }
+};
+
+// create a change json if the old and new value are different, otherwise return undefined
+export const createChangeJsonNonList = (
+  nameOfField: string,
+  oldValue: any,
+  newValue: any,
+  crId: number,
+  implementerId: number,
+  wbsElementId: number
+) => {
+  if (oldValue == null) {
+    return {
+      changeRequestId: crId,
+      implementerId,
+      wbsElementId,
+      detail: `Added ${nameOfField} "${newValue}"`
+    };
+  } else if (oldValue !== newValue) {
+    return {
+      changeRequestId: crId,
+      implementerId,
+      wbsElementId,
+      detail: `Edited ${nameOfField} from "${oldValue}" to "${newValue}"`
+    };
+  }
+  return undefined;
+};
+
+// create a change json list for a given list (rules). Only works if the elements themselves should be compared (strings)
+export const createRulesChangesJson = (
+  nameOfField: string,
+  oldArray: string[],
+  newArray: string[],
+  crId: number,
+  implementerId: number,
+  wbsElementId: number
+) => {
+  const seenOld = new Set<string>(oldArray);
+  const seenNew = new Set<string>(newArray);
+
+  const changes: { element: string; type: string }[] = [];
+
+  oldArray.forEach((element) => {
+    if (!seenNew.has(element)) {
+      changes.push({ element, type: 'Removed' });
+    }
+  });
+
+  newArray.forEach((element) => {
+    if (!seenOld.has(element)) {
+      changes.push({ element, type: 'Added new' });
+    }
+  });
+
+  return changes.map((element) => {
+    return {
+      changeRequestId: crId,
+      implementerId,
+      wbsElementId,
+      detail: `${element.type} ${nameOfField} "${element.element}"`
+    };
+  });
+};
+
+// this method creates changes for description bullet inputs
+// it returns it as an object of {deletedIds[], addedDetails[] changes[]}
+// because the deletedIds are needed for the database and the addedDetails are needed to make new ones
+export const createDescriptionBulletChangesJson = (
```

Comment:
This is the method that is giving me trouble. It originally would list any description bullet that had been deleted as deleted again. I fixed that by filtering out the deleted ones. Now it lists any that don't change as deleted and then added again.